### ISSUE TITLE
ttnn.slice accepts H/W end bounds past the logical shape and returns tile padding

### ISSUE THREAD

**Issue by kanapitsas** [tenstorrent/tt-metal#55471]
## Summary

For `TILE_LAYOUT` tensors, `ttnn.slice` accepts H/W end bounds past the logical shape as
long as they remain within the tile-rounded padded shape. The extra output contains the
implicit tile padding as logical data.

This violates the documented contract that `slice_end[i]` must be in
`(0, input_tensor_shape[i]]`. `Tensor.__getitem__` uses the same path:
`x[:, :, 32:64, :]` on 40 rows returns 32 rows instead of torch's 8.

## Observed

Tensor `(1, 1, 40, 40)`, `TILE_LAYOUT`, padded shape `(1, 1, 64, 64)`:

| dim | requested end | logical | result |
| --- | --- | --- | --- |
| 0 | 2 | 1 | **rejected** — `Ends 2 must be less than or equal to the shape of the tensor 1` |
| 1 | 2 | 1 | **rejected** |
| 2 (height) | 48, then 64 | 40 | **accepted**, returns padding |
| 3 (width) | 48, then 64 | 40 | **accepted**, returns padding |
| 2 or 3 | 65 (past the padding) | 40 | rejected |

The excess region tracks `pad_value` exactly, ruling out stale memory or noise:

| `from_torch(..., pad_value=v)` | values returned past the logical shape |
| --- | --- |
| 0.0 | 0.0 |
| 7.0 | 7.0 |
| -3.0 | -3.0 |
| 123.0 | 123.0 |

## Repro

```python
import torch
import ttnn

d = ttnn.open_device(device_id=0)
t = torch.arange(1 * 1 * 40 * 40, dtype=torch.float32).reshape(1, 1, 40, 40) + 1.0

x = ttnn.from_torch(t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=d)
print("logical", tuple(t.shape), "padded", x.padded_shape)
out = ttnn.to_torch(ttnn.slice(x, [0, 0, 0, 0], [1, 1, 40, 64]))
print("slice to width 64 ->", tuple(out.shape), "  torch would give", tuple(t[:, :, :, :64].shape))

# Positive controls: the same over-large bound is rejected on dims 0 and 1.
for ends, dim in (([2, 1, 40, 40], 0), ([1, 2, 40, 40], 1)):
    try:
        ttnn.slice(x, [0, 0, 0, 0], ends)
        print(f"dim {dim}: accepted (unexpected)")
    except Exception as e:
        print(f"dim {dim}: rejected ->", str(e).splitlines()[0][:80])

for v in (0.0, 7.0, -3.0, 123.0):
    y = ttnn.from_torch(t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=d, pad_value=v)
    got = ttnn.to_torch(ttnn.slice(y, [0, 0, 0, 0], [1, 1, 40, 64]))
    print(f"pad_value={v:>7} -> excess region [{got[..., 40:].min():.1f}, {got[..., 40:].max():.1f}]")

print("x[:, :, 32:64, :] ->", tuple(ttnn.to_torch(x[:, :, 32:64, :]).shape),
      "  torch:", tuple(t[:, :, 32:64, :].shape))

ttnn.close_device(d)
```

Observed output: direct slice `(1, 1, 40, 64)` vs torch `(1, 1, 40, 40)`; both controls
rejected; every excess region equal to its `pad_value`; Python indexing
`(1, 1, 32, 40)` vs torch `(1, 1, 8, 40)`, including **960 padding elements**.

## Root cause

`slice.cpp` starts from `input_tensor.logical_shape()` and normalizes signed bounds with:

```cpp
modified_ends[i] = operations::data_movement::wrap_index(ends[i], input_shape[i]);
```

and `wrap_index` (`ttnn/cpp/ttnn/operations/data_movement/common/common.cpp`) is

```cpp
uint32_t wrap_index(int index, int size) { return index < 0 ? size + index : index; }
```

`wrap_index` only handles negative indices; no positive upper-bound check against the
logical shape follows. For tiled input, `slice.cpp` rounds the last two `modified_ends`
into `padded_ends`, then the device operation checks them against `input.padded_shape()`.
Leading dimensions reject because their padded and logical extents match; H/W accept ends
up to the tile-rounded extent.

## Suggested fix

Validate every direct-API end against `logical_shape[i]` before constructing `padded_ends`;
the existing error text can be reused. For `Tensor.__getitem__`, clamp the stop to the
logical extent before calling `ttnn.slice`, preserving Python/torch semantics.

## Related: #53284

Open issue #53284 reports that `prod_bw` passes `padded_shape()` bounds to `ttnn::slice`.
This remains silent because `slice` accepts them; logical validation would fail immediately.

I also tested that path. For accepted shapes `(1,1,32,40)`, `(1,1,32,48)`,
`(1,1,64,40)`, and `(1,1,32,33)` with `dim=3`, gradients match torch autograd within
bfloat16 quantization; other tested combinations fail an upstream broadcast check. This
report is about slice validation and padding exposure, not a demonstrated `prod_bw`
miscomputation.

## Environment

- Reproduced on tt-metal `main` @ `7632d6f67ceb2826d5caf5e33881aa07da3e3803`.
- Still present in source on `main` @ `172d28fa8ed6adea36dce09bf94687941fb34968`
  (2026-09-04).
- Executed on the **ttsim functional simulator** (Blackhole), not silicon; I do not have a
  card. Acceptance and output metadata are host-side. The value check only identifies
  explicit `pad_value` contents; no device-arithmetic claim is made.

## Prior art checked

Re-checked on 2026-09-04 with `ttnn.slice in:title` and broader searches for slice padding,
logical shapes, `padded_shape`, and `wrap_index`. I reviewed #10170, #16857, #26119,
#42778, #46954, #51214, #51279, and #53284; none reports this missing logical upper bound.

#10170 is the closest title match, but used an aligned `(96, 32)` tensor and in-range slices.
It reported output padding metadata and closed as fixed in January 2025. This report uses
an unaligned shape and out-of-range end that expose input padding on current `main`.

**Comment by kanapitsas** [comment]
I have the reproducer and a targeted fix scoped, including regression tests for both
`ttnn.slice` and tensor indexing. Could you assign this issue to me?

**Comment by jberkowitzTT** [comment]
Closing this out while we review it for the bounty program
